hbase-mapreduce/src/main/java/org/apache/hadoop/hbase/mapreduce/replication/VerifyReplicationRecompareRunnable.java

```diff
@@ -0,0 +1,156 @@
+
+  private final int reCompareTries;
+  private final int sleepMsBeforeReCompare;
+  private final int reCompareBackoffExponent;
+  private final boolean verbose;
+
+  private Result sourceResult;
+  private Result replicatedResult;
+
+  public VerifyReplicationRecompareRunnable(Mapper.Context context, Result sourceResult,
+    Result replicatedResult, VerifyReplication.Verifier.Counters originalCounter, String delimiter,
+    Scan tableScan, Table sourceTable, Table replicatedTable, int reCompareTries,
+    int sleepMsBeforeReCompare, int reCompareBackoffExponent, boolean verbose) {
+    this.context = context;
+    this.sourceResult = sourceResult;
+    this.replicatedResult = replicatedResult;
+    this.originalCounter = originalCounter;
+    this.delimiter = delimiter;
+    this.tableScan = tableScan;
+    this.sourceTable = sourceTable;
+    this.replicatedTable = replicatedTable;
+    this.reCompareTries = reCompareTries;
+    this.sleepMsBeforeReCompare = sleepMsBeforeReCompare;
+    this.reCompareBackoffExponent = reCompareBackoffExponent;
+    this.verbose = verbose;
+    this.row = VerifyReplication.getRow(sourceResult, replicatedResult);
+  }
+
+  @Override
+  public void run() {
+    Get get = new Get(row);
+    get.setCacheBlocks(tableScan.getCacheBlocks());
+    get.setFilter(tableScan.getFilter());
+
+    int sleepMs = sleepMsBeforeReCompare;
+    int tries = 0;
+
+    while (++tries <= reCompareTries) {
+      context.getCounter(VerifyReplication.Verifier.Counters.RE_COMPARES).increment(1);
+
+      try {
+        Thread.sleep(sleepMs);
+      } catch (InterruptedException e) {
+        LOG.warn("Sleeping interrupted, incrementing bad rows and aborting");
+        incrementOriginalAndBadCounter();
+        Thread.currentThread().interrupt();
+      }
+
+      try {
+        if (fetchLatestRows(get)) {
+          if (matches(sourceResult, replicatedResult, null)) {
+            if (verbose) {
+              LOG.info("Good row key (with recompare): {}{}{}", delimiter,
+                Bytes.toStringBinary(row), delimiter);
+            }
+            context.getCounter(VerifyReplication.Verifier.Counters.GOODROWS).increment(1);
+            return;
+          }
+        }
+      } catch (IOException e) {
+        context.getCounter(VerifyReplication.Verifier.Counters.FAILED_RECOMPARE).increment(1);
+        if (verbose) {
+          LOG.info("Got an exception during recompare for rowkey={}", Bytes.toStringBinary(row), e);
+        }
+      }
+
+      sleepMs = sleepMs * (2 ^ reCompareBackoffExponent);
+    }
+
+    LOG.error("{}, rowkey={}{}{}", originalCounter, delimiter, Bytes.toStringBinary(row),
+      delimiter);
+    incrementOriginalAndBadCounter();
+  }
+
+  private boolean fetchLatestRows(Get get) throws IOException {
+    Result sourceResult = sourceTable.get(get);
+    Result replicatedResult = replicatedTable.get(get);
+
+    boolean sourceMatches = matches(sourceResult, this.sourceResult,
```

Comment:
I made a small change here which always executes the `matches` method for both results. This ensures we capture both `SOURCE_ROW_CHANGED` and `PEER_ROW_CHANGED` instead of just `SOURCE_ROW_CHANGED` when applicable.